### ISSUE TITLE
Restriction conflict in strings

### ISSUE THREAD

**Issue by mbaluda** [mulesoft/api-workbench#453]
The attached example does not pass the validation while it should according to the RAML spec.
In fact, example c is accepted by both the constraints of his type and supertype[]
[patterns.txt](https://github.com/mulesoft/api-workbench/files/909794/patterns.txt)

**Comment by ddenisenko** [comment]
Its not an example passing the validation, its the `String2` subtype not passing it. According to the spec it is not allowed to re-define pattern in subtypes:

```
If a sub-type inherits properties having the same name from at least two of its parent types, the sub-type keeps all restrictions applied to those properties with two exceptions: 1) a "pattern" facet when a parent type already declares a "pattern" facet 2) a user-defined facet when another user-defined facet has the same value. In these cases, an invalid type declaration occurs.
```

I agree the error message should be more descriptive.

**Comment by mbaluda** [comment]
The sentence you are quoting is related to multiple inheritance.
The raml-org/raml-java-parser validates both regexp in such case

**Comment by ddenisenko** [comment]
>The sentence you are quoting is related to multiple inheritance.
Agree. The issue is, this is the closest statement from the spec that I could find regarding this case.

If you can find another one better describing it, please do. 

@petrochenko-pavel-a @sichvoge Guys, what do you think should happen if there are two string types, one extends another, and both define built-in `pattern` facet?

**Comment by mbaluda** [comment]
The standard is not specific about this case and therefore I expect consistency with other restrictions like minLength and maxLength.
Such restrictions define a subset of valid values. Subtypes need to satisfy all the restrictions in the hierarchy therefore only accept the intersection of all the subsets of valid values.

Consider the attached example where the correctly implemented behavior of minLength and maxLength is highlighted
[RestrictionInheritance.txt](https://github.com/mulesoft/api-workbench/files/917352/RestrictionInheritance.txt)

**Comment by KevinMitchell** [comment]
In the case of things like length you could argue that subtypes should only be allowed to make a constraint more restrictive.  So c-string in the example should be rejected because it's trying to increase the max length past the limit defined in a-string.  So ideally we shouldn't just take the intersection of the constraints in the case of inheritances chains like this, as it perhaps masks a problem we should be notified about.  If we really want to take the intersection of independent constraints then we should probably create a new type that inherits from both of them IMO.

**Comment by mbaluda** [comment]
That would be subsumption, which is a perfectly valid semantics for subtyping but it's not what I see implemented in the available RAML processors.
I'll check how popular XSD and Jsonschema processors behave.

**Comment by KevinMitchell** [comment]
I realise that's not the behaviour currently implemented in some of the processors.  But if the spec needs firming up in this area anyway then it should presumably define the desired behaviour, whatever that might be, rather than just formalise what the RAML parser(s) currently do?
